clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,119 @@
+// RUN: %clang_cc1 %s -fsyntax-only -ast-dump -fsycl-is-device -triple spir64 -verify
+
+// expected-error 4 @+2{{kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) {
+  kernelFunc();
+}
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct T {};
+
+template <typename Func>
+void kernel3func(const Func &F3) {
+  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  constexpr const char *F3_output = __builtin_unique_stable_name(F3);
+  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
```

Comment:
@erichkeane -- do you think the diagnostic you emit should have some way of identifying which instantiation is the issue? There are two situations I'm worried this may be unclear:
```
kernel_single_task<SomethingPotentiallyBad>(SomeElsePotentiallyBad); // Which name is the issue?
kernel_with_multiple_args<class Whatever>(SomethingPotentiallyBad, SomethingElsePotentiallyBad); // Which name is the issue?
```
Also, the diagnostic should put single quotes around `__builtin_unique_stable_name`.

Reply:
I don't think a kernel can have multiple functors, so I don't see much value there (that is, I don't think the 2nd line is valid).  Additionally, based on how we're implemented, I don't think we really know WHICH name is the problem.  The mechanism is to store the old name, re-evaluate it and see if we've changed it at all, so the best we know is the builtin call that changes, and the kernel call that caused it.

  As for the single-quote, thats a good feedback.  I've just pushed that change to my repo.